protocols/gossipsub/src/handler.rs

```diff
@@ -571,10 +538,32 @@ impl ConnectionHandler for Handler {
             ConnectionEvent::FullyNegotiatedOutbound(fully_negotiated_outbound) => {
                 self.on_fully_negotiated_outbound(fully_negotiated_outbound)
             }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError { error: e, .. }) => {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
                 self.outbound_substream_establishing = false;
-                warn!("Dial upgrade error {:?}", e);
-                self.upgrade_errors.push_back(e);
+
+                match error {
+                    // Timeout errors get mapped to NegotiationTimeout and we close the connection.
+                    ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer => {
+                        self.keep_alive = KeepAlive::No;
+                        log::info!("Dial upgrade error: Protocol negotiation timeout.");
+                    }
+                    // There was an error post negotiation, close the connection.
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)) => {
+                        log::info!("Dial upgrade error: {e}");
+                    }
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(negotiation_error)) => {
+                        match negotiation_error {
+                            NegotiationError::Failed => {
+                                // The protocol is not supported
+                                self.protocol_unsupported = true;
+                                log::info!("Dial upgrade error: {}", NegotiationError::Failed);
+                            }
+                            NegotiationError::ProtocolError(e) => {
+                                log::info!("Gossipsub error: Protocol negotiation failed. {e}");
+                            }
+                        }
```

Comment:
You can merge this match statement into the outer one and thus reduce one level of indentation.

Reply:
done